src/activity/activity_heartbeat_manager.rs

```diff
@@ -0,0 +1,86 @@
+use crate::errors::ActivityHeartbeatError;
+use crate::protos::coresdk::{common, ActivityHeartbeat};
+use crate::ServerGatewayApis;
+use crossbeam::channel::{bounded, unbounded, Receiver, Select, Sender, TryRecvError};
+use dashmap::{DashMap, DashSet};
+use futures::FutureExt;
+use std::convert::TryInto;
+use std::future::Future;
+use std::ops::Div;
+use std::sync::Arc;
+use std::{thread, time};
+use tokio::time::{sleep, Sleep};
+
+pub(crate) struct ActivityHeartbeatManager {
+    /// Core will aggregate activity heartbeats and send them to the server periodically.
+    /// This map contains latest heartbeat details for each activity that reported them.
+    heartbeat_details: DashMap<Vec<u8>, Vec<common::Payload>>,
+
+    scheduled_heartbeats: DashSet<Vec<u8>>,
+
+    heartbeat_tx: Sender<ActivityHeartbeatTask>,
+
+    shutdown_tx: Sender<bool>,
+}
+
+struct ActivityHeartbeatTask {
+    task_token: Vec<u8>,
+    next_delay: time::Duration,
+    trigger: Sleep,
+}
+
+impl ActivityHeartbeatManager {
+    pub fn new() -> Self {
+        let (heartbeat_tx, heartbeat_rx) = unbounded::<ActivityHeartbeatTask>();
+        let (shutdown_tx, shutdown_rx) = bounded(1);
+        thread::spawn(move || Self::processor(heartbeat_rx, shutdown_rx));
+        Self {
+            heartbeat_details: Default::default(),
+            scheduled_heartbeats: Default::default(),
+            heartbeat_tx,
+            shutdown_tx,
+        }
+    }
+
+    pub fn record(&self, heartbeat: ActivityHeartbeat) -> Result<(), ActivityHeartbeatError> {
+        self.heartbeat_details
+            .insert(heartbeat.task_token.clone(), heartbeat.details);
+
+        if self
+            .scheduled_heartbeats
+            .insert(heartbeat.task_token.clone())
+        {
+            let initial_delay = time::Duration::from_millis(0);
+            let heartbeat_timeout: time::Duration = heartbeat
+                .heartbeat_timeout
+                .ok_or(ActivityHeartbeatError::HeartbeatTimeoutNotSet)?
+                .try_into()
+                .or(Err(ActivityHeartbeatError::InvalidHeartbeatTimeout))?;
+            let next_delay = heartbeat_timeout.div(2);
+
+            self.heartbeat_tx
+                .send(ActivityHeartbeatTask {
+                    task_token: heartbeat.task_token,
+                    next_delay,
+                    trigger: sleep(initial_delay),
+                })
+                .expect("failed to send activity task into the channel");
+        };
+
+        Ok(())
+    }
+
+    pub fn shutdown(&self) {
+        self.shutdown_tx.send(true);
+    }
+
+    fn processor(heartbeat_rx: Receiver<ActivityHeartbeatTask>, shutdown_rx: Receiver<bool>) {
+        // TODO this function should:
+        // 1. consume heartbeat tasks
+        // 2. join the futures and for any future that becomes ready:
```

Comment:
This operation is called select, not join. Joining them all means they all must finish, whereas you only care about when one finishes. https://docs.rs/futures/0.3.14/futures/future/fn.select_all.html would do what you want.

However, I think there is a simpler design here.

There is no need for the channels and this separate processor thread/task I think. The design can instead be to keep one map of token -> spawn handle, where the spawn handle is the result of a future you have spawned whose job is to issue heartbeats for that task token (like an actor model). Then instead of having this processor thread that manages all hearbeat tasks, you have a task-per-task-token with a simpler responsibility (just managing the timing of sending out heartbeats for an individual activity). Upon shutdown, you just join all the spawn handles, ensuring final heartbeats are sent.

The per-activity spawned task would (mostly pseudocode) look something like:
```rust
let channel_for_activity_tx, channel_for_activity_rx = tokio::sync::watch()
// Put tx side in the map with the spawn handle
async {
  let timeout_or_watch_changed = select!(sleep(no_heartbeats_duration), channel_for_activity_rx.changed());
  while(timeout_or_watch_changed.await != timed out) {
      send_heartbeat_to_server(...).await;
      sleep(interval between sending batches).await;
  }
}
```

Reply:
The watch channel can contain the heartbeat details

Reply:
Thanks Spencer, let me do it this way.